// artisan-delegator/lib/errors.js
var util = require('util')
  ;

function NotFoundError(message) {
  this.message = message || 'Not Found';
  Error.captureStackTrace(this, this);
}

util.inherits(NotFoundError, Error);

NotFoundError.prototype.name = 'NotFoundError';

NotFoundError.prototype.code = 404;

exports.NotFoundError = NotFoundError;

function MethodNotAllowedError(message) {
  this.message = message || 'Method Not Allowed';
  Error.captureStackTrace(this, this);
}

util.inherits(MethodNotAllowedError, Error);

exports.MethodNotAllowedError = MethodNotAllowedError;

MethodNotAllowedError.prototype.name = 'MethodNotAllowedError';

MethodNotAllowedError.prototype.code = 405;

function ParserError(message) {
  this.message = message || 'Parser Error';
  Error.captureStackTrace(this, this);
}

util.inherits(ParserError, Error);

exports.ParserError = ParserError;

ParserError.prototype.name = 'ParserError';


function MappingFileError(message) {
  this.message = message || 'Mapping File Error';
  Error.captureStackTrace(this, this);
}

util.inherits(MappingFileError, Error);

exports.MappingFileError = MappingFileError;

MappingFileError.prototype.name = 'MappingFileError';
